lib/explorer/shared.ex

```diff
@@ -215,32 +224,85 @@ defmodule Explorer.Shared do
   defp type(item, _type) when is_binary(item), do: :string
 
   defp type(item, _type) when is_nil(item), do: nil
+  defp type([], _type), do: nil
+  defp type([_item | _] = items, type), do: {:list, result_list_type(items, type)}
   defp type(item, _type), do: raise(ArgumentError, "unsupported datatype: #{inspect(item)}")
 
+  defp result_list_type(nil, _type), do: nil
+  defp result_list_type([], _type), do: nil
+
+  defp result_list_type([h | _tail] = items, type) when is_list(h) do
+    # Enum.flat_map/2 is used here becase we want to remove one level of nesting per iteraction.
+    {:list, result_list_type(Enum.flat_map(items, & &1), type)}
+  end
+
+  defp result_list_type(items, type) when is_list(items) do
+    check_types!(items, leaf_dtype(type))
+  end
+
+  defp leaf_dtype({:list, inner_dtype}), do: leaf_dtype(inner_dtype)
+  defp leaf_dtype(dtype), do: dtype
+
   @doc """
   Downcasts lists of mixed numeric types (float and int) to float.
   """
   def cast_numerics(list, type) when type == :numeric do
-    data =
-      Enum.map(list, fn
-        item when item in [nil, :infinity, :neg_infinity, :nan] -> item
-        item -> item / 1
-      end)
+    {cast_numerics_to_floats(list), :float}
+  end
 
-    {data, :float}
+  def cast_numerics(list, {:list, _} = dtype) do
+    {cast_numerics_deep(list, dtype), cast_numeric_dtype_to_float(dtype)}
   end
 
   def cast_numerics(list, type), do: {list, type}
 
+  defp cast_numerics_to_floats(list) do
+    Enum.map(list, fn
+      item when item in [nil, :infinity, :neg_infinity, :nan] or is_float(item) -> item
+      item -> item / 1
+    end)
+  end
+
+  defp cast_numerics_deep(nil, _), do: nil
+
+  defp cast_numerics_deep(list, {:list, inner_dtype}) when is_list(list) do
+    Enum.map(list, fn item -> cast_numerics_deep(item, inner_dtype) end)
+  end
+
+  defp cast_numerics_deep(list, :numeric), do: cast_numerics_to_floats(list)
+  defp cast_numerics_deep(list, _), do: list
+
+  defp cast_numeric_dtype_to_float({:list, :numeric}), do: {:list, :float}
+  defp cast_numeric_dtype_to_float({:list, other} = dtype) when is_atom(other), do: dtype
+
+  defp cast_numeric_dtype_to_float({:list, inner}),
+    do: {:list, cast_numeric_dtype_to_float(inner)}
+
+  defp cast_numeric_dtype_to_float(other), do: other
+
+  @doc """
+  Check if the dtype has the leaf in one of the possible values.
+
+  This is useful to verify if the list dtype of unknown "height" is valid,
+  or if it has the dtype in one of the possibilities.
+  """
+  def leaf_dtype_of?({:list, inner_dtype}, possibilities) do
+    leaf_dtype_of?(inner_dtype, possibilities)
+  end
```

Comment:
Suggestion: drop this function and make `leaf_dtype` public. This function duplicates the logic of `leaf_dtype`, it's roughly the same number of characters to do `if leaf_dtype(dtype) in list` as it is `if leaf_dtype_of?(dtype, list)`.

Reply:
Good call! :D
I made the changes in 695a948147744b76a77745596ff7fdd32b179478